File: backend/app/schemas/sse.py
from __future__ import annotations

import uuid
from typing import Annotated, Literal

from pydantic import BaseModel, Field

from app.db.models.game import GameStatus
from app.schemas.play import PlayRead
from app.schemas.shap import ShapFeature


class PlayUpdateEvent(BaseModel):
    event_type: Literal["play_update"] = "play_update"
    game_id: str
    play: PlayRead
    home_wp: float
    away_wp: float
    top_shap: list[ShapFeature]


class GameStatusEvent(BaseModel):
    event_type: Literal["game_status"] = "game_status"
    game_id: str
    status: GameStatus
    home_score: int
    away_score: int


class ReplayCompleteEvent(BaseModel):
    event_type: Literal["replay_complete"] = "replay_complete"
    game_id: str


SSEEvent = Annotated[
    PlayUpdateEvent | GameStatusEvent | ReplayCompleteEvent,
    Field(discriminator="event_type"),
]

File: frontend/src/types/sse.ts
import type { Play } from './play'
import type { ShapFeature } from './prediction'
import type { GameStatus } from './game'

export interface PlayUpdateEvent {
  event_type: 'play_update'
  game_id: string
  play: Play
  home_wp: number
  away_wp: number
  top_shap: ShapFeature[]
}

export interface GameStatusEvent {
  event_type: 'game_status'
  game_id: string
  status: GameStatus
  home_score: number
  away_score: number
}

export interface ReplayCompleteEvent {
  event_type: 'replay_complete'
  game_id: string
}

export type SSEEvent = PlayUpdateEvent | GameStatusEvent | ReplayCompleteEvent
